### PR TITLE
Code editor

### PR DESCRIPTION
**Pull request by eadm**
**YouTrack task**: [#APPS-1530](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1530)

**Description List**:
* Add CodeEditor view that supports code highlighting:
    * Code parsing performed by [Java port](https://code.google.com/archive/p/java-prettify/) of [Google prettify JS library](https://github.com/google/code-prettify)
    * Following languages are supported: C, C++, C#, Java, JS, Rust,  CoffeeScript, Perl, Python, Ruby, Sh, Appollo, Basic, Clojure, CSS, Dart, Erlang, Go, Haskell, Kotlin, Lisp, Scala, TeX, Matlab, R, SQL, Yaml, Wiki, Lua, Ml, Mumps, N, Pascal, Proto, Rd
    * Color themes support
* Add RxJava & RxAndroid to project

### CODE REVIEW

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    var lang = ""
+        set(value) {
+            field = value
+            afterTextChanged(editableText) // refresh highlight
+        }
+
+    
+    var scrollContainer: ScrollView? = null
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: what if we will set this value more than 1 time? ScrollPublisher will publish events to old observer (memory leak)

Comment (KirillMakarov) [reply app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: We need to dispose old observer (if it is existed) and after that subscribe new

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        )
+
+        compositeDisposable.add(
+                highlightPublisher
+                        .debounce(INPUT_DEBOUNCE_MS, TimeUnit.MILLISECONDS)
+                        .map {
+                            parser.parse(lang, it.toString())
+                        }
+                        .subscribeOn(Schedulers.computation())
+                        .observeOn(Schedulers.computation())
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: `observeOn/subscribeOn` is useless here

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+            lineNumbersStrokePaint.color = value.lineNumberStroke
+            lineNumbersTextPaint.color = value.lineNumberText
+            selectedLinePaint.color = value.selectedLineBackground
+
+            afterTextChanged(editableText) // refresh highlight
+        }
+
+    private val bufferRect = Rect()
+
+    var lines: List<String> = emptyList()
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: Can we make it `private` and `val`?

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    var linesWithNumbers: List<Int> = emptyList()
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: Can we make it `private` and `val`?

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+    override fun onDraw(canvas: Canvas) {
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: `onDraw` allocates a lot of memory now. Please avoid it. http://prntscr.com/gwtu61

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        val lineNumbersOffset = lineNumbersTextPaint.measureText(lineCount.toString()).toInt() + 2 * LINE_NUMBERS_MARGIN
+
+        setPadding(lineNumbersOffset + LINE_NUMBERS_MARGIN, paddingTop, paddingRight, paddingBottom)
+        canvas.drawRect(0f, 0f, lineNumbersOffset.toFloat(), height.toFloat(), lineNumbersBackgroundPaint) // line numbers bg
+        canvas.drawLine(lineNumbersOffset.toFloat(), 0f, lineNumbersOffset.toFloat(), height.toFloat(), lineNumbersStrokePaint) // line numbers stroke
+
+        layout?.let { layout ->
+            var pos = 0
+            linesWithNumbers = lines.map {
+                val line = layout.getLineForOffset(pos)
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: allocation. can we precalc it?

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+    private val bufferRect = Rect()
+
+    var lines: List<String> = emptyList()
+    var linesWithNumbers: List<Int> = emptyList()
+
+    override fun onDraw(canvas: Canvas) {
+        val lineNumbersOffset = lineNumbersTextPaint.measureText(lineCount.toString()).toInt() + 2 * LINE_NUMBERS_MARGIN
+
+        setPadding(lineNumbersOffset + LINE_NUMBERS_MARGIN, paddingTop, paddingRight, paddingBottom)
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: why left padding is different?

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+            }
+
+            val cursorPosition = selectionStart
+            if (cursorPosition > 0) {
+                val selectedLine = layout.getLineForOffset(cursorPosition)
+                var lineStart = selectedLine
+                var lineEnd = selectedLine + 1
+
+                while (!linesWithNumbers.contains(lineStart) && lineStart > 0) lineStart--
+                while (!linesWithNumbers.contains(lineEnd) && lineEnd < lineCount) lineEnd++
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: what do this code? Please, extract it to the function. Can we optimize it (use hashset for example)?

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt (hunk 1 above)
Comment (KirillMakarov) [reply app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: Why we should set `scrollContainer`, if edit text is scrollable? Check `Scroller` class

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt (hunk 8 above)
Comment (eadm) [reply app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: This code looks for upper and lover bound of text line. Usually text lines don't take more than 5 editor lines so there is no need in additional structures.

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt (hunk 7 above)
Comment (eadm) [reply app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: To create space for line numbers.

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt (hunk 4 above)
Comment (eadm) [reply app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: This values will be used in CodeAnalyzer for smart brackets and etc

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt (hunk 3 above)
Comment (eadm) [reply app/src/main/java/org/stepic/droid/code/ui/CodeEditor.kt]: This values will be used in `CodeAnalyzer` for smart brackets and etc

File: app/src/main/java/org/stepic/droid/code/ui/CodeEditorLayout.kt [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+            setBackgroundColor(value.background)
+            codeEditor.theme = value
+        }
+
+    val text : CharSequence
+        get() = codeEditor.text.toString()
+
+    constructor(context: Context) : this(context, null)
+
+    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/code/ui/CodeEditorLayout.kt]: We can use @JvmOverload and default args in constructor (see `LoadingView`). I think it is more elegant solution for custom views

File: app/src/main/java/org/stepic/droid/ui/fragments/CodeStepFragment.kt [hunk 10]
```diff
@@ -74,7 +73,7 @@ class CodeStepFragment : StepAttemptFragment(),
         codeToolbarAdapter?.onSymbolClickListener = this
         codeToolbarView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
 
-        codeEditor.minLines = MIN_LINES_IN_ANSWER_FIELD
+//        codeEditor.minLines = MIN_LINES_IN_ANSWER_FIELD
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/fragments/CodeStepFragment.kt]: please, remove commented code.

File: app/src/main/res/layout/code_editor.xml [hunk 11]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<merge>
```
Comment (KirillMakarov) [inline app/src/main/res/layout/code_editor.xml]: Is merge tag is really needed here?

**Pull request merged**